keps/sig-auth/3766-referential-authorization/README.md

```diff
@@ -413,80 +429,121 @@ rejected".
-//
-// If any of the above occur, a generic error message such as "RefNotPermitted" 
-// should be communicated, likely via status on the referring resource.
-//
-// Support: Core
-type ReferenceGrant struct {
-	metav1.TypeMeta   `json:",inline"`
-	metav1.ObjectMeta `json:"metadata,omitempty"`
+// Subject refers to the subject that is a consumer of the referenced
+// pattern(s).
+type ClusterReferenceConsumer struct {
+    metav1.TypeMeta   `json:",inline"`
+    metav1.ObjectMeta `json:"metadata,omitempty"`
+
+    // Subject refers to the subject that is a consumer of the referenced
+    // pattern(s).
+    Subject Subject       `json:"subject"`
+
+    // From refers to the group and resource that these references originate from.
+    From    GroupResource `json:"from"`
+
+    // To refers to the group and resource that these references target.
+    To      GroupResource `json:"to"`
+
+    // For refers to the purpose of this reference. (Cluster)ReferenceGrants
+    // matching the From, To, and For of this resource will be authorized for
+    // the Subject of this resource.
+    //
+    // This value must be a valid DNS label as defined per RFC-1035.
+    For     string        `json:"for"`
+}
 
-	// Spec defines the desired state of ReferenceGrant.
-	Spec ReferenceGrantSpec `json:"spec,omitempty"`
+// Subject is a copy of RBAC Subject that excludes APIGroup.
+type Subject struct {
+    // Kind of object being referenced. Values defined by this API group are
+    // "User", "Group", and "ServiceAccount". If the Authorizer does not
+    // recognized the kind value, the Authorizer should report an error.
+    Kind string `json:"kind"`
+    // Name of the object being referenced.
+    Name string `json:"name"`
+    // Namespace of the referenced object.  If the object kind is non-namespace,
+    // such as "User" or "Group", and this value is not empty the Authorizer
+    // should report an error. +optional
+    Namespace string `json:"namespace,omitempty"`
+}
+```
+
+#### ClusterReferenceGrant
 
-	// Note that `Status` sub-resource has been excluded at the
-	// moment as it was difficult to work out the design.
-	// A `Status` sub-resource may be added in the future.
+```golang
+// ClusterReferenceGrant authorizes same-namespace references matching the
+// specified pattern.
+type ClusterReferenceGrant struct {
+    metav1.TypeMeta   `json:",inline"`
+    metav1.ObjectMeta `json:"metadata,omitempty"`
+
+    // From refers to the group, resource, and pthat that these references
+    // originate from.
+    From    GroupVersionResourcePath `json:"from"`
+
+    // To refers to the group and resource that these references target.
+    To      GroupResource `json:"to"`
+
+    // For refers to the purpose of this reference. Subjects of
+    // ClusterReferenceConsumers will be authorized to follow references matching
+    // the From, To, and For of this resource.
+    //
+    // This value must be a valid DNS label as defined per RFC-1035.
```

Comment:
```suggestion
    // This value must be a valid DNS label as defined per RFC-1035, and is matched case-sensitively.
```

Reply:
Can we have this as a case-insenstive match?

Reply:
Yes, we can, I just want to clarify whether we have one or the other. As long as it's documented, I'm fine with either.

Reply:
The relevant RFC actually only supports lowercase characters, will update the KEP to clarify: https://github.com/kubernetes/kubernetes/blob/d2904845cdbbe16f418a3352ae4312775e657d24/pkg/apis/apiserverinternal/validation/validation.go#L180-L181